vdr/src/types/transaction.rs

```diff
@@ -495,3 +495,336 @@ pub struct BlockDetails {
+                None,
+                CONTRACT_METHOD_EXAMPLE,
+                Some(TransactionType::Read),
+                None,
+            )
+            .await
+            .unwrap_err();
+
+            assert!(matches!(err, VdrError::ContractInvalidName { .. }));
+        }
+
+        #[async_std::test]
+        async fn build_txn_contract_method_does_not_set() {
+            let err = setup_transaction_builder(
+                CONTRACT_NAME_EXAMPLE,
+                None,
+                Some(TransactionType::Read),
+                None,
+            )
+            .await
+            .unwrap_err();
+
+            assert!(matches!(err, VdrError::ContractInvalidName { .. }));
+        }
+
+        #[async_std::test]
+        async fn build_txn_contract_method_does_not_exist() {
+            let err = setup_transaction_builder(
+                CONTRACT_NAME_EXAMPLE,
+                INVALID_METHOD,
+                Some(TransactionType::Read),
+                None,
+            )
+            .await
+            .unwrap_err();
+
+            assert!(matches!(err, VdrError::ContractInvalidName { .. }));
+        }
+
+        #[async_std::test]
+        async fn build_txn_write_sender_does_not_set() {
+            let err = setup_transaction_builder(
+                CONTRACT_NAME_EXAMPLE,
+                CONTRACT_METHOD_EXAMPLE,
+                Some(TransactionType::Write),
+                None,
+            )
+            .await
+            .unwrap_err();
+
+            assert!(matches!(err, VdrError::ClientInvalidTransaction { .. }));
+        }
+
+        #[async_std::test]
+        #[ignore]
+        async fn build_txn_invalid_from_address() {
+            let err = setup_transaction_builder(
+                CONTRACT_NAME_EXAMPLE,
+                CONTRACT_METHOD_EXAMPLE,
+                Some(TransactionType::Write),
+                INVALID_ACC_ADDRESS,
+            )
+            .await
+            .unwrap_err();
+
+            assert!(matches!(err, VdrError::ClientInvalidTransaction { .. }));
+        }
+
+        #[async_std::test]
+        async fn build_txn_valid_transactions_with_0x_prefix() {
+            setup_transaction_builder(
+                CONTRACT_NAME_EXAMPLE,
+                CONTRACT_METHOD_EXAMPLE,
+                Some(TransactionType::Write),
+                Some(&Address::default().to_string()),
+            )
+            .await
+            .unwrap();
+        }
+
+        #[async_std::test]
+        async fn build_txn_valid_transactions_without_0x_prefix() {
+            let address_without_prefix = VALIDATOR_ADDRESS.trim_start_matches("0x");
+            setup_transaction_builder(
+                CONTRACT_NAME_EXAMPLE,
+                CONTRACT_METHOD_EXAMPLE,
+                Some(TransactionType::Write),
+                Some(address_without_prefix),
+            )
+            .await
+            .unwrap();
+        }
+
+        #[async_std::test]
```

Comment:
it would be much better to use rstest for all these test

Reply:
used rstest